### PR TITLE
Proposal: Add generic to setState method

### PR DESCRIPTION
**Pull request by tombarton**
Apologies to keep raising PRs, I really don't want to be _that_ person!

This is merely a proposal, so feel free to shoot it down/tear it apart. I'm in the process of creating custom layers using TS and I'm trying to keep things as strict as possible. When extending a layer class, I'm able to type the state appropriately, however, I lose type safety when it comes to the `setState` method. I'm proposing adding a generic to the `setState` method to facilitate this. Examples below:

**Strictly Typed:**

```
class CustomLayer<D, P extends Props<D>> extends CompositeLayer<D, P> {
    state!: CustomLayerState<D>;

    updateState(...args) {
        // This is all just pseudo code
        this.setState<CustomLayerState<D>>({
            foo: "bar"
        })
    }
}
```

**Loosely Typed:**

```
class CustomLayer extends CompositeLayer<any> {
    updateState(...args) {
        this.setState({
            foo: "bar"
        })
    }
}
```

This would give people the opportunity to pursue strict typing if they so choose, but wouldn't affect anyone who is currently using `setState` as it currently is.

### CODE REVIEW

File: deck.gl__core/index.d.ts [hunk 1]
```diff
@@ -1280,8 +1280,6 @@ declare module "@deck.gl/core/lib/composite-layer" {
 		get isComposite(): boolean;
 		get isLoaded(): any;
 		getSubLayers(): any;
-		initializeState(params?: any): void;
```
Comment (tombarton) [inline deck.gl__core/index.d.ts:1283]: `CompositeLayer` extends `Layer`, therefore these types should not be required. I've tested this locally, but please let me know if I've overlooked something.

**Comment by danmarshall** [comment]
> Apologies to keep raising PRs, I really don't want to be that person!
Quite the opposite, we appreciate all the help! Keep the code coming 👍

File: deck.gl__core/index.d.ts (hunk 1 above)
Comment (danmarshall) [reply deck.gl__core/index.d.ts:1283]: as long as it builds. Can you run `npm test` locally without error?

Comment (tombarton) [reply deck.gl__core/index.d.ts:1283]: Had a few issues with tests in the end due to a recent release of `luma.gl` including types which were conflicting with those in this library. As a result, I've had to specify versions of luma that do not include said types. This is a dev dependency so it won't have any impact on end users, however, I've raised an issue (#137) to look at resolving this properly.

**Comment by danmarshall** [comment]
Please see #147, which fixes #137, which hopefully unblocks this.

**Comment by danmarshall** [comment]
@tombarton thanks for the contribution, unfortunately I think it's out of date. If you want to resolve the merge conflict and re-open this, please feel free. Thanks

**Pull request closed without merging**